Handle rejected promises inside incoming Invocation messages

damirault opened the pull request:
# Handle rejected promises inside incoming Invocation messages 

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->
## Description
Incoming messages of Invocation type result in executing a clientMethod promise, but promise rejections are unhandled.  This can result in a unhandled rejected promise, potentially resulting in Node.js process crashing. Note, this was previously detected via eslint and suppressed in code. The user-code callback itself is already [covered by a try-catch block](https://github.com/damirault/aspnetcore/blob/64c3630930f14bbf3b585410a81f0bab739d5ba6/src/SignalR/clients/ts/signalr/src/HubConnection.ts#L787), it's the wrapper invocation function and the sending of the response message that can trigger the promise rejection.

To avoid this, catch promise rejections and log the failure.

One scenario this case can happen is, if during the clientMethod call, the server connection is disconnected, the invokation will still attempt to send a completion message (through _sendWithProtocol). This will then result in a promise rejection, such as the following:

```
Cannot send data if the connection is not in the 'Connected' State.
    at HttpConnection.send (node_modules\@microsoft\signalr\dist\cjs\HttpConnection.js:95:35)
    at HubConnection._sendMessage (node_modules\@microsoft\signalr\dist\cjs\HubConnection.js:266:32)
    at HubConnection._sendWithProtocol (node_modules\@microsoft\signalr\dist\cjs\HubConnection.js:273:21)
    at HubConnection._invokeClientMethod (node_modules\@microsoft\signalr\dist\cjs\HubConnection.js:577:24)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Fixes #52524

damirault committed:
Handle rejected promises inside incoming Invocation messages

Incoming messages of Invocation type result in executing a clientMethod promise, but promise rejections are unhandled.
This can result in a unhandled rejected promise, potentially resulting in Node.js process crashing.
Note, this was previously detected via eslint and suppressed in code.

To avoid this, catch promise rejections and log the failure.

One scenario this case can happen is, if during the clientMethod call, the server connection is disconnected, the invokation will still attempt to send a completion message (through _sendWithProtocol). This will then result in a promise rejection, such as the following:

```
Cannot send data if the connection is not in the 'Connected' State.
    at HttpConnection.send (node_modules\@microsoft\signalr\dist\cjs\HttpConnection.js:95:35)
    at HubConnection._sendMessage (node_modules\@microsoft\signalr\dist\cjs\HubConnection.js:266:32)
    at HubConnection._sendWithProtocol (node_modules\@microsoft\signalr\dist\cjs\HubConnection.js:273:21)
    at HubConnection._invokeClientMethod (node_modules\@microsoft\signalr\dist\cjs\HubConnection.js:577:24)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

halter73 commented:
> Note, this was previously detected via eslint and suppressed in code.

I think we suppressed this because its library code calling into user code. I suppose we could make this clearer with the comment, but what if someone already has this in their process to handle such a case?

```js
process.on('unhandledRejection', (reason, promise) => {
  console.log('Unhandled Rejection at:', promise, 'reason:', reason);
  // Application specific logging, throwing an error, or other logic here
});
```

https://nodejs.org/api/process.html#event-unhandledrejection
https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event

This change could break the app by swallowing and logging rejection rather than triggering the 'unhandledRejection' event. I think it should be up to the application developer to handle any rejected promises in their client methods since SignalR has no idea whether the app has special handlers configured.

Maybe I could be convinced the current behavior is too much of a footgun and SignalR should handle these anyway, since eslint will not notice the promise is unhandled since it's returned and then suppressed. But then, I think we would need some mechanism to observer the error. Possibly something like `HubConnection.onerror`. The old jQuery-based SignalR client used to have this, but we've avoided it so far with the new client and haven't received many complaints until now. At least, not that I'm aware of.

halter73 commented:
> One scenario this case can happen is, if during the clientMethod call, the server connection is disconnected, the invokation will still attempt to send a completion message (through _sendWithProtocol). This will then result in a promise rejection, such as the following:

I should have read that part more carefully the first time. We already swallow and log client callback rejections [here](
https://github.com/damirault/aspnetcore/blob/690d26ab4e1fde992119ade0bc8c7da10b79a87b/src/SignalR/clients/ts/signalr/src/HubConnection.ts#L791).

I'm guessing the reason we haven't seen this before is that it requires one of the client callbacks to actually be async for us to attempt to send the completion message without `_processIncomingData` still on the synchronous stack. And as long as we're still processing incoming data synchronously, we would remain in the connected state and never see this error. Can we add a regression test for this in https://github.com/damirault/aspnetcore/blob/64c3630930f14bbf3b585410a81f0bab739d5ba6/src/SignalR/clients/ts/signalr/tests/HubConnection.test.ts? We already register an unhandledRejection handler that should fail the test.

https://github.com/dotnet/aspnetcore/blob/0787f9fa52bc822a83e2e3e73a272bd18d754e9a/src/SignalR/clients/ts/signalr/tests/HubConnection.test.ts#L24

damirault committed:
Add test of callback sending response with a rejected promise

Adds a test covering when callback invoked when server invokes a method on the client and then handles rejected promise on send

When the unhandled promise fix is not applied to HubConnection,
 the following error fails the HubConnection.test.ts test suite:

    Send error

      831 |                         connection.send = () => {
      832 |                             promiseRejected = true;
    > 833 |                             return Promise.reject(new Error("Send error"));
          |                                                   ^
      834 |                         }
      835 |                         p.resolve();
      836 |                     });

      at TestConnection.connection.send (signalr/tests/HubConnection.test.ts:833:51)
      at HubConnection.send [as _sendMessage] (signalr/src/HubConnection.ts:422:32)
      at HubConnection._sendMessage [as _sendWithProtocol] (signalr/src/HubConnection.ts:433:25)
      at HubConnection._sendWithProtocol [as _invokeClientMethod] (signalr/src/HubConnection.ts:808:24)

damirault commented:
Makes sense. Test case added.

halter73 reviewed (approved):
Thank you! I want to give @BrennanConroy a chance to take a look before merging. Would we consider backporting this?

BrennanConroy commented:
/backport to release/8.0
